### PR TITLE
feat: rewrite Elasticsearch index and search code to achieve Elasticsearch vector and full-text search

### PR DESCRIPTION
**Pull request by KennFalcon**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. Close issue syntax: `Fixes #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

Fixes 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: api/core/rag/datasource/vdb/elasticsearch/elasticsearch_vector.py [hunk 1]
```diff
@@ -37,12 +43,19 @@ class ElasticSearchVector(BaseVector):
     def __init__(self, index_name: str, config: ElasticSearchConfig, attributes: list):
         super().__init__(index_name.lower())
         self._client = self._init_client(config)
+        self._version = self._get_version()
+        self._check_version()
```
Comment (bowenliang123) [inline api/core/rag/datasource/vdb/elasticsearch/elasticsearch_vector.py:47]: The `_version` is used only once. Consider nest the `_get_version` method inside `_check_version` without holding it as member variable of each ElasticSearchVector instance.

**Comment by KennFalcon** [comment]
Because ES 8.15.0 has a parameter k in the knn query statement in the query section, which 8.14.3 does not have, the test did not pass. Currently, I have modified the code to a knn query method that is compatible with both version 8.14.0 and 8.15.0.

File: api/configs/middleware/__init__.py [hunk 2]
```diff
@@ -200,5 +201,6 @@ class MiddlewareConfig(
     TencentVectorDBConfig,
     TiDBVectorConfig,
     WeaviateConfig,
+    ElasticsearchConfig
```
Comment (bowenliang123) [inline api/configs/middleware/__init__.py]:
```suggestion
    ElasticsearchConfig,
```

Place `ElasticsearchConfig ` after `ChromaConfig ` aligning the alphabet order.
And use a comma separator.

**Comment by bowenliang123** [comment]
> Because ES 8.15.0 has a parameter k in the knn query statement in the query section, which 8.14.3 does not have, the test did not pass. Currently, I have modified the code to a knn query method that is compatible with both version 8.14.0 and 8.15.0.

Yes, knn query is an important feature for ES vector. Are you testing it against both 8.14.x and 8.15.+? Otherwise, they are never well maintained.

File: api/core/rag/datasource/vdb/elasticsearch/elasticsearch_vector.py [hunk 3]
```diff
@@ -53,42 +66,27 @@ def _init_client(self, config: ElasticSearchConfig) -> Elasticsearch:
 
         return client
 
+    def _get_version(self) -> str:
+        info = self._client.info()
```
Comment (bowenliang123) [inline api/core/rag/datasource/vdb/elasticsearch/elasticsearch_vector.py:70]: Is this a client-side version only?

Comment (KennFalcon) [reply api/core/rag/datasource/vdb/elasticsearch/elasticsearch_vector.py:70]: No, this return cluster information.
image

**Comment by KennFalcon** [comment]
> > Because ES 8.15.0 has a parameter k in the knn query statement in the query section, which 8.14.3 does not have, the test did not pass. Currently, I have modified the code to a knn query method that is compatible with both version 8.14.0 and 8.15.0.
> 
> Yes, knn query is an important feature for ES vector. Are you testing it against both 8.14.x and 8.15.+? Otherwise, they are never well maintained.

In the current ES, there are two way use knn query.
1. https://www.elastic.co/guide/en/elasticsearch/reference/8.15/query-dsl-knn-query.html
2. https://www.elastic.co/guide/en/elasticsearch/reference/8.15/knn-search.html

The second way is the way that has been supported since 8.0.0（https://www.elastic.co/guide/en/elasticsearch/reference/8.0/knn-search.html）, so the current code has been changed to this way. But the lower version of ES only supports l2_norm, not cosine. So may need to do a compatibility later.

**Comment by wellCh4n** [comment]
I formatted a file and now the CI is all done~

**Review by JohnJyong — APPROVED** [review]

**Pull request merged**